src/include/ipc4/header.h

```diff
@@ -0,0 +1,103 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2021 Intel Corporation. All rights reserved.
+ */
+
+/**
+ * \file include/ipc4/header.h
+ * \brief IPC4 gloabl definitions
+ */
+
+#ifndef __SOF_IPC4_HEADER_H__
+#define __SOF_IPC4_HEADER_H__
+
+#include <stdint.h>
+
+//! Message target, value of msg_tgt field.
+enum ipc4_message_target {
+	//! Global FW message
+	SOF_IPC4_MESSAGE_TARGET_FW_GEN_MSG = 0,
+	//! Module message
+	SOF_IPC4_MESSAGE_TARGET_MODULE_MSG = 1
+};
+
+//! Message direction, value of rsp field.
+enum ipc4_message_direction {
+	//! Request, Notification
+	SOF_IPC4_MESSAGE_DIR_MSG_REQUEST = 0,
+	//! Reply
+	SOF_IPC4_MESSAGE_DIR_MSG_REPLY = 1,
+};
+
+/*
+ * Global IPC4 message types - must fit into 5 bits.
+ */
+enum ipc4_message_type {
+	//! Boot Config.
+	SOF_IPC4_GLB_BOOT_CONFIG = 0,
+	//! ROM Control (directed to ROM).
+	SOF_IPC4_GLB_ROM_CONTROL = 1,
+	//! Execute IPC gateway command
+	SOF_IPC4_GLB_IPCGATEWAY_CMD = 2,
+	//! Create EDF task and run FreeRTOS instance in it
+	SOF_IPC4_GLB_START_RTOS_EDF_TASK = 3,
+	//! Stop FreeRTOS and delete its EDF task context
+	SOF_IPC4_GLB_STOP_RTOS_EDF_TASK = 4,
+
+	/* GAP size 8 here - can 5 .. 12  be used ?? */
+
+	//! Execute performance measurements command.
+	SOF_IPC4_GLB_PERF_MEASUREMENTS_CMD = 13,
+	//! DMA Chain command.
+	SOF_IPC4_GLB_CHAIN_DMA = 14,
+	//! Load multiple modules
+	SOF_IPC4_GLB_LOAD_MULTIPLE_MODULES = 15,
+	//! Unload multiple modules
+	SOF_IPC4_GLB_UNLOAD_MULTIPLE_MODULES = 16,
+	//! Create pipeline
+	SOF_IPC4_GLB_CREATE_PIPELINE = 17,
+	//! Delete pipeline
+	SOF_IPC4_GLB_DELETE_PIPELINE = 18,
+	//! Set pipeline state
+	SOF_IPC4_GLB_SET_PIPELINE_STATE = 19,
+	//! Get pipeline state
+	SOF_IPC4_GLB_GET_PIPELINE_STATE = 20,
+	//! Get pipeline context size
+	SOF_IPC4_GLB_GET_PIPELINE_CONTEXT_SIZE = 21,
+	//! Save pipeline
+	SOF_IPC4_GLB_SAVE_PIPELINE = 22,
+	//! Restore pipeline
+	SOF_IPC4_GLB_RESTORE_PIPELINE = 23,
+	//! Loads library (using Code Load or HD/A Host Output DMA)
+	SOF_IPC4_GLB_LOAD_LIBRARY = 24,
+	//! Internal FW message
+	SOF_IPC4_GLB_INTERNAL_MESSAGE = 26,
+	//! Notification (FW to SW driver)
+	SOF_IPC4_GLB_NOTIFICATION = 27,
+
+	/* Gap size 3 here - can be 28 .. 30 used ?? */
+
+	//! Maximum message number
+	SOF_IPC4_GLB_MAX_IXC_MESSAGE_TYPE = 31
+};
+
+union ipc4_message_header {
+	uint32_t dat;
+
+	struct {
+		uint32_t rsvd0		: 24;
+
+		//! One of Global::Type
+		uint32_t type		: 5;
+
+		//! Msg::MSG_REQUEST
+		uint32_t rsp		: 1;
+
+		//! Msg::FW_GEN_MSG
+		uint32_t msg_tgt	: 1;
+
+		uint32_t _reserved_0: 1;
```

Comment:
NAK. bitfields are not portable and compiler-dependent. we've been there before.

Reply:
let's go back to traditional way.

Reply:
Completely agree, but this is an exact copy of the interface (converted from C++)  and the host middleware wont work if it's changed. This is a one time and one device only interface. 

See https://github.com/thesofproject/ipc-tuples for future work direction (still WIP).

Reply:
@lgirdwood I'm just getting back to working on SOF again (sorry for my absence)  but I agree with @plbossart and @RanderWang here in that the bit fields are completely non portable and compiler dependent. They should never be used for generic ABIs across different architectures.

Reply:
First I used marco definition to access each filed , but I found redundancy between macro definition and bit field, then I deleted macro definition since bit filed is more friendly to programmer. https://github.com/RanderWang/sof/blob/3a4b7476c619cba5b774f4f77d561fd55c6a52e4/src/include/ipc/header2.h#L78-L80 .  So we can go back.

Reply:
@jgunn0262 I fully agree, bit fields are compiler dependent and it's just good fortune that in this case they are aligned between the FW and host compilers. Yes you are right, this is not guaranteed to work in future compilers.
@RanderWang We cant upstream the bit fields to Linux (since there are many maintainers would would probably raise issues), but I think given the other comments, we should be able to **copy exactly** into firmware as long as we contain/restrict this code.

Reply:
@lgirdwood  I don't use bit filed in kernel.  you meant copy kernel header file to firmware ?

Reply:
@RanderWang I dont want to copy to firmware if we dont have to, there are a lot of headers that use these "base" IPC structure and that would mean updating them all.

Reply:
@lgirdwood It's not great upstreaming the bit fields, but I think you are saying this is for a single Intel device only ? 

I've no problem if I can still use the existing IPC or the ipc-tuples which looks portable, but I'm still looking at the tuples. Can you give me some reassurances here and are there any other IPC4 headers you can share ? 

@dbaluta I see that you don't work on Intel hardware so I'm interested what you think about this IPC version ?

Reply:
@jgunn0262 yes, this is for one Intel device only atm. More headers are here https://github.com/thesofproject/sof/tree/lrg/topic/ipc4/src/include/ipc4 (branch is rebased frequently)

Reply:
@jgunn0262 agree, the ABI is quite Intel specific and 
1) it will be wholly enclosed within the src/ipc directory (with abstraction where we do IPC communication outside of this area).
2) it will be build time selected in Kconfig.
This would mean none of your code would depend on it at either runtime or build time and you can continue with the current IPC.
Does this work you you ?

Reply:
@lgirdwood yes, this should be fine.